musig/README.md

```diff
@@ -34,19 +34,63 @@ Ristretto base point in compressed form:
+  suitable challenge for that public key. 
+- `get_pubkeys(&self) -> Vec<VerificationKey>`: returns the associate public keys.
+
+### Multikey (implements MusigContext)
 
 Fields:
 - transcript: `Transcript`. All of the pubkeys that the multikey are created from are committed to this transcript. 
 - aggregated_key: `VerificationKey`
+- public_keys: `Vec<VerificationKey`
 
 Functions: 
-- `Multikey::new(...)`: detailed more in [key aggregation](#key-aggregation) section. 
-- `Multikey::factor_for_key(&self, &verification_key)`: computes the `a_i` factor, where `a_i = H(<L>, X_i)`. 
-  `<L>`, or the list of pukeys that go into the aggregated pubkey, has already been committed into `self.transcript`. Therefore, this function simply clones `self.transcript`, commits the verification key (`X_i`) into the transcript 
-    with label "X_i", and then squeezes the challenge scalar `a_i` from the transcript with label "a_i".
+- `Multikey::new(...) -> Self`: detailed more in [key aggregation](#key-aggregation) section. 
+
+- `Multikey::commit(&self, &mut transcript)`: Commits `self.aggregated_key` to the input `transcript` with label "X".
+
+- `Multikey::challenge(&self, &verification_key, &mut transcript, &nonce_commitment) -> Scalar`: 
+  Computes challenge `c_i = a_i * c`, where `a_i = H_agg(<L>, X_i)` and `c = H_sig(X, R, m)`.
+
+  For calculating `a_i`, `<L>` (the list of pubkeys that go into the aggregated pubkey)
+  has already been committed into `self.transcript`. Therefore this function simply clones `self.transcript`, 
+  commits the verification key (`X_i`) into the transcript with label "X_i", 
+  and then squeezes the challenge scalar `a_i` from the transcript with label "a_i".
+
+  For calculating `c`, the message `m` has already been committed to the input `transcript`, 
+  so the function only needs to commit `self.aggregated_key` with label "X" and input `hash_commitment` with label "R", 
+  and then get the challenge scalar `c` with label "c".
+
 - `Multikey::aggregated_key(&self)`: returns the aggregated key stored in the multikey, `self.aggregated_key`.  
 
+- `Multikey::get_pubkeys(&self)`: returns the list of public keys, `self.public_keys`.
+
+### Multimessage (implements MusigContext)
+
+Fields:
+- pairs: `Vec<(VerificationKey, [u8])`
+
+Functions:
+- `Multimessage::new(Vec<(VerificationKey, [u8])) -> Self`: creates a new MultiMessage instance using the inputs.
+
+- `Multimessage::commit(&self, &mut transcript)`: It commits each of the pairs in `self.pairs` to the input `transcript`,
+  by iterating through `self.pairs` and committing the `VerificationKey` with label "X" and the message with label "m".
+
+- `Multimessage::challenge(&self, &verification_key, &mut transcript, &nonce_commitment) -> Scalar`: 
+  Computes challenge `c_i = H(R, <S>, i)`.
+  It first commits the input `nonce_commitment` to the input `transcript` with label "R".
```

Comment:
Per notes above, the R is assumed to be committed by the caller (Party::receive_commitments), and pairs are committed by the caller too via Multimessage::commit().

Here we only need to figure out the `i`, fork the transcript, commit `i`, get challenge, and then in the non-forked transcript add domain separation with `commit("dom-sep", "Musig.multi-message-boundary")`.

Does this make sense?

Reply:
it doesn't seem to be a huge difference whether we commit `R` and the context in the `MusigContext::challenge` method (as I detailed in the first draft of the writeup), or outside it in the `PartyAwaitingCommitments::receive_commitments` method (as you are suggesting). The only thing I would be cautious of with your proposal, is making sure that the transcript has actually had `R` and the context committed to it before squeezing the challenge! (Whereas with the first draft, you don't have to worry about it because the method commits those for you). 

I'll change the writeup to your suggestion!